utils/postgis_geocoder.py

```diff
@@ -0,0 +1,194 @@
+        # Create a temporary values table to JOIN to the area table (multiple times)
+        parameters = []
+        values = []
+        for i, coordinate in enumerate(coordinates):
+            values.append("(%s, %s, %s, ST_SetSRID(ST_MakePoint(%s, %s), 4326))")
+            parameters.append(i)
+            parameters.append(coordinate.x)
+            parameters.append(coordinate.y)
+            parameters.append(coordinate.x)
+            parameters.append(coordinate.y)
+
+        values = f"VALUES {', '.join(values)}"
+
+        # For each area type, add a SELECT and a LEFT JOIN
+        area_selects = []
+        area_joins = []
+
+        # Area type format is:
+        # (alias, LIH code/MapIt codes, MapIt generation)
+        area_types = [
+            ("postcode", "PC", None),
+            ("postcode_sector", "PCS", None),
+            ("postcode_district", "PCD", None),
+            ("postcode_area", "PCA", None),
+            ("output_area", "OA21", None),
+            ("lsoa", "LSOA", None),
+            ("msoa", "MSOA", None),
+            ("wmc", "WMC23", CURRENT_MAPIT_GENERATION),
+            ("ward", "WD23", CURRENT_MAPIT_GENERATION),
+            (
+                "county",
+                ["CTY"],
+                CURRENT_MAPIT_GENERATION,
+            ),  # Array is interpreted as MapIt code instead of AreaType code
+            (
+                "district",
+                ["LBO", "UTA", "COI", "LGD", "MTD", "DIS", "NMD"],
+                CURRENT_MAPIT_GENERATION,
+            ),
+            ("eer", "EER", None),
+        ]
+        for alias, code, mapit_gen in area_types:
+            area_selects.append(
+                f"{alias}.gss AS {alias}_code, {alias}.name AS {alias}_name"
+            )
+
+            if isinstance(code, str):
+                area_filter = f"area_type_id = (SELECT id FROM hub_areatype WHERE code = '{code}')"
+            else:
+                codes = [f"'{c}'" for c in code]
+                area_filter = f"mapit_type IN ({', '.join(codes)})"
+
+            if mapit_gen:
+                area_filter += f" AND mapit_generation_high = {mapit_gen}"
+
+            gis_query = f"SELECT id FROM hub_area WHERE {area_filter} ORDER BY temp.point <-> polygon LIMIT 1"
+            join = f"LEFT JOIN hub_area AS {alias} ON {alias}.id = ({gis_query})"
+            area_joins.append(join)
+
+        query = f"""
+        SELECT
+          temp.i,
+          temp.y AS latitude,
+          temp.x AS longitude,
+          {',\n'.join(area_selects)}
+        FROM
+          (
+            SELECT * FROM ({values}) AS temp (i, x, y, point)
+          ) AS temp
+          {'\n'.join(area_joins)}
+        """
+
+        cursor.execute(query, parameters)
+        rows = cursor.fetchall()
+        columns: list[str] = [col[0] for col in cursor.description]
+
+    # Combine the results by input index, keeping the areas with the highest MapIt generation
+    # (not sure how to do this in the database query)
```

Comment:
Yeah it's finnicky, probably some kind of window function / subquery sort by limit 1 manoeuvre. But if this is fast enough then whatever.